dbms/src/Storages/Page/V3/CheckpointFile/tests/gtest_file_read_write.cpp

```diff
@@ -0,0 +1,634 @@
+                },
+            },
+        });
+        edits.appendRecord({.type = EditRecordType::VAR_REF, .page_id = "foo", .ori_page_id = "abc"});
+        edits.appendRecord({.type = EditRecordType::VAR_ENTRY, .page_id = "aaabbb", .entry = {.offset = 10}});
+        edits.appendRecord({.type = EditRecordType::VAR_DELETE, .page_id = "sun"});
+        writer->writeEditsAndApplyRemoteInfo(edits);
+    }
+    writer->writeSuffix();
+    writer.reset();
+
+    auto manifest_reader = CPManifestFileReader::create({
+        .file_path = dir + "/manifest_foo",
+    });
+    manifest_reader->readPrefix();
+    CheckpointProto::StringsInternMap im;
+    {
+        auto edits_r = manifest_reader->readEdits(im);
+        auto r = edits_r->getRecords();
+        ASSERT_EQ(4, r.size());
+
+        ASSERT_EQ(EditRecordType::VAR_ENTRY, r[0].type);
+        ASSERT_EQ("abc", r[0].page_id);
+        ASSERT_EQ(0, r[0].entry.offset); // The deserialized offset is not the same as the original one!
+        ASSERT_EQ(0, r[0].entry.size);
+        ASSERT_TRUE(r[0].entry.checkpoint_info->is_local_data_reclaimed); // After deserialization, this field is always true!
+        ASSERT_EQ("my_file_id", *r[0].entry.checkpoint_info->data_location.data_file_id);
+
+        ASSERT_EQ(EditRecordType::VAR_REF, r[1].type);
+        ASSERT_EQ("foo", r[1].page_id);
+        ASSERT_EQ("abc", r[1].ori_page_id);
+
+        ASSERT_EQ(EditRecordType::VAR_ENTRY, r[2].type);
+        ASSERT_EQ("aaabbb", r[2].page_id);
+        ASSERT_EQ(0, r[2].entry.offset);
+        ASSERT_EQ(0, r[2].entry.size);
+        ASSERT_TRUE(r[2].entry.checkpoint_info->is_local_data_reclaimed);
+        ASSERT_EQ("data_1", *r[2].entry.checkpoint_info->data_location.data_file_id);
+        ASSERT_EQ("nahida opened her eyes", readData(r[2].entry.checkpoint_info->data_location));
+
+        ASSERT_EQ(EditRecordType::VAR_DELETE, r[3].type);
+        ASSERT_EQ("sun", r[3].page_id);
+    }
+    {
+        auto edits_r = manifest_reader->readEdits(im);
+        ASSERT_FALSE(edits_r.has_value());
+    }
+    {
+        auto locks = manifest_reader->readLocks();
+        ASSERT_TRUE(locks.has_value());
+        ASSERT_EQ(2, locks->size());
+        ASSERT_EQ(1, locks->count("data_1"));
+        ASSERT_EQ(1, locks->count("my_file_id"));
+    }
+    {
+        auto locks = manifest_reader->readLocks();
+        ASSERT_FALSE(locks.has_value());
+    }
+}
+CATCH
+
+TEST_F(CheckpointFileTest, FromBlobStore)
+try
+{
+    const auto delegator = std::make_shared<DB::tests::MockDiskDelegatorMulti>(std::vector{dir});
+    const auto file_provider = DB::tests::TiFlashTestEnv::getContext().getFileProvider();
+    auto blob_store = BlobStore<universal::BlobStoreTrait>(getCurrentTestName(), file_provider, delegator, BlobConfig{});
+
+    auto edits = universal::PageEntriesEdit{};
+    {
+        UniversalWriteBatch wb;
+        wb.putPage("page_foo", 0, "The flower carriage rocked");
```

Comment:
How about add a page with `data_sizes`, and make sure the `fields_offset` is correct from manifest file?

Reply:
Done. Please take a look again, thanks!